core/rawdb/accessors_l1_message.go

```diff
@@ -0,0 +1,209 @@
+	batch := db.NewBatch()
+	WriteL1Messages(batch, l1Msgs)
+	if err := batch.Write(); err != nil {
+		log.Crit("Failed to store L1 message batch", "err", err)
+	}
+}
+
+// ReadL1MessageRLP retrieves an L1 message in its raw RLP database encoding.
+func ReadL1MessageRLP(db ethdb.Reader, enqueueIndex uint64) rlp.RawValue {
+	data, err := db.Get(L1MessageKey(enqueueIndex))
+	if err != nil {
+		log.Crit("Failed to load L1 message", "enqueueIndex", enqueueIndex, "err", err)
+	}
+	return data
+}
+
+// ReadL1Message retrieves the L1 message corresponding to the enqueue index.
+func ReadL1Message(db ethdb.Reader, enqueueIndex uint64) *types.L1MessageTx {
+	data := ReadL1MessageRLP(db, enqueueIndex)
+	if len(data) == 0 {
+		return nil
+	}
+	l1Msg := new(types.L1MessageTx)
+	if err := rlp.Decode(bytes.NewReader(data), l1Msg); err != nil {
+		log.Crit("Invalid L1 message RLP", "enqueueIndex", enqueueIndex, "data", data, "err", err)
+	}
+	return l1Msg
+}
+
+// L1MessageIterator is a wrapper around ethdb.Iterator that
+// allows us to iterate over L1 messages in the database. It
+// implements an interface similar to ethdb.Iterator.
+type L1MessageIterator struct {
+	inner     ethdb.Iterator
+	keyLength int
+}
+
+// IterateL1MessagesFrom creates an L1MessageIterator that iterates
+// over all L1 message in DB starting at the provided enqueue index.
+func IterateL1MessagesFrom(db ethdb.Iteratee, fromEnqueueIndex uint64) L1MessageIterator {
+	start := encodeEnqueueIndex(fromEnqueueIndex)
+	it := db.NewIterator(L1MessagePrefix, start)
+	keyLength := len(L1MessagePrefix) + 8
+
+	return L1MessageIterator{
+		inner:     it,
+		keyLength: keyLength,
+	}
+}
+
+// Next moves the iterator to the next key/value pair.
+// It returns whether the iterator is exhausted.
+func (it *L1MessageIterator) Next() bool {
+	for it.inner.Next() {
+		key := it.inner.Key()
+		if len(key) == it.keyLength {
+			return true
+		}
+	}
+	return false
+}
+
+// EnqueueIndex returns the enqueue index of the current L1 message.
+func (it *L1MessageIterator) EnqueueIndex() uint64 {
+	key := it.inner.Key()
+	raw := key[len(L1MessagePrefix) : len(L1MessagePrefix)+8]
+	enqueueIndex := binary.BigEndian.Uint64(raw)
+	return enqueueIndex
+}
+
+// L1Message returns the current L1 message.
+func (it *L1MessageIterator) L1Message() types.L1MessageTx {
+	data := it.inner.Value()
+	l1Msg := types.L1MessageTx{}
+	if err := rlp.DecodeBytes(data, &l1Msg); err != nil {
+		log.Crit("Invalid L1 message RLP", "data", data, "err", err)
+	}
+	return l1Msg
+}
+
+// Release releases the associated resources.
+func (it *L1MessageIterator) Release() {
+	it.inner.Release()
+}
+
+// ReadL1MessagesInRange retrieves all L1 messages between two enqueue indices (inclusive).
+// The resulting array is ordered by the L1 message enqueue index.
+func ReadL1MessagesInRange(db ethdb.Iteratee, firstEnqueueIndex, lastEnqueueIndex uint64, checkRange bool) []types.L1MessageTx {
+	expectedCount := lastEnqueueIndex - firstEnqueueIndex + 1
```

Comment:
I think here is needed check that `firstEnqueueIndex` < `lastEnqueueIndex`, otherwise following `make` may fail